### PR TITLE
[SMTChecker] Fix push().push()

### LINKED ISSUE #9021: [SMTChecker] ICE because array push not supported on dynamically sized arrays within structs
## Description

```
pragma experimental SMTChecker;
contract C {
  struct S {
    int[] b;
  }
  S s;
  function f() public { s.b.push(); }
}
```

throws

https://github.com/ethereum/solidity/blob/370350da07a158f0a3a68efc5e1f65583e19b57e/libsolidity/formal/SMTEncoder.cpp#L1150

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ solc <input.sol>
```

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/ethereum/solidity/issues/9021

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ Bugfixes:
  * SMTChecker: Fix internal error when using bitwise operators on fixed bytes type.
  * SMTChecker: Fix internal error when using compound bitwise operator assignments on array indices inside branches.
  * SMTChecker: Fix error in events with indices of type static array.
+ * SMTChecker: Fix internal error in sequential storage array pushes (``push().push()`).
```
Comment (chriseth) [inline Changelog.md:15]:
```suggestion
 * SMTChecker: Fix internal error in sequential storage array pushes (``push().push()``).
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 2]
```diff
@@ -1119,6 +1119,28 @@ void SMTEncoder::arrayPushPopAssign(Expression const& _expr, smtutil::Expression
...
+		{
+			auto memberAccess = dynamic_cast<MemberAccess const*>(&funCall->expression());
+			solAssert(memberAccess, "");
+			auto symbArray = dynamic_pointer_cast<smt::SymbolicArrayVariable>(m_context.expression(memberAccess->expression()));
+			solAssert(symbArray, "");
+
+			auto oldLength = symbArray->length();
+			auto store = smtutil::Expression::store(
+				symbArray->elements(),
+				symbArray->length() - 1,
```
Comment (chriseth) [inline libsolidity/formal/SMTEncoder.cpp:1135]: Shouldn't this be just `length()`?

File: libsolidity/formal/SMTEncoder.cpp [hunk 3]
```diff
@@ -1119,6 +1119,28 @@ void SMTEncoder::arrayPushPopAssign(Expression const& _expr, smtutil::Expression
...
+				_array
+			);
+			symbArray->increaseIndex();
+			m_context.addAssertion(symbArray->elements() == store);
+			m_context.addAssertion(symbArray->length() == oldLength);
```
Comment (chriseth) [inline libsolidity/formal/SMTEncoder.cpp:1140]: Shouldn't this be `oldLength + 1`?

File: libsolidity/formal/SMTEncoder.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:1135]: One annoying part of this implementation is that in `array.push().push()` the inner `array.push()` will run first, so it's length will already be adjusted. Here we only assign the last item to the result of the right `push()`.

File: libsolidity/formal/SMTEncoder.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:1140]: Because of the above

File: libsolidity/formal/SMTEncoder.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:1135]: The code you're referring to is above in `void SMTEncoder::arrayPush(FunctionCall const& _funCall)`. The code is very similar, so I'm still trying to find a way to merge those pieces of code and avoid repetition.

Comment (chriseth) [reply libsolidity/formal/SMTEncoder.cpp:1135]: This looks to me like a special case that should be solved in a more general way. What if you have something like
```
function f(... x) internal returns (...) { return x.push(); }
function g() internal {
  f(...).push();
}
```

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:1135]: Yea also things like
```
uint[][] a;
function f(uint x) internal view returns (uint[] storage) { returns a[x]; }
function g(uint x) public { f(x).push(); }
```
wouldn't be handled by that.
The problem with these 2 cases is that storage pointers erase knowledge anyway

**Review by chriseth — APPROVED** [review]

**Pull request merged**